### ISSUE TITLE
Support for injecting passwords from Kubernetes secrets into config files

### ISSUE THREAD

**Issue by gburton1** [splunk/splunk-operator#657]
Is there support for injecting a password to a config file from a Kubernetes secret? For example, in authentication.conf, you might have LDAP bind creds:
```
[Active Directory]
bindDN = ldapuser.example.net
bindDNpassword = <reference to Kubernetes secret>
```
I see the concept of password management covers some important passwords, but it doesn't seem to cover the general case of any sensitive content in any config file.
https://github.com/splunk/splunk-operator/blob/master/docs/PasswordManagement.md#password-management

**Comment by pogdin** [comment]
Thanks for the feedback.  We definitely see this in other places as well, namely in passwords.conf that need to be populated in Splunk Apps.

**Comment by locmai** [comment]
Just to clarify my understanding, these kind of `.conf` files are configured via the `CommonSplunkSpec.Defaults` that shared across the other CRDs:
https://github.com/splunk/splunk-operator/blob/6a46c20f22e1c86a22872e00a6c28f526c07610b/pkg/apis/enterprise/v2/common_types.go#L88

And reconcile it into a configmap via:
https://github.com/splunk/splunk-operator/blob/6a46c20f22e1c86a22872e00a6c28f526c07610b/pkg/splunk/enterprise/util.go#L124-L131

So we could either:
- Add a Secret reference in the CommonSplunkSpec then template those secret's values later inside the configmap value in ApplySplunkConfig func
- Change Defaults from a normal string attribute to an object and do the same.

**Comment by locmai** [comment]
My very naive attempt: https://github.com/locmai/splunk-operator/commit/524757f13c8cc480dc0a4e90c97a0baec6dc885c

**Comment by gburton1** [comment]
@pogdin we might be able to help with this by contributing in the near future. If you know of any design direction that has already been decided for how to handle this case, or you have a particular way in mind, please let us know. As you can see above, Loc Mai already started playing with some code changes just to get the conversation started.

**Comment by pogdin** [comment]
@locmai and @gburton1, thanks again for your interest in the project.  @akondur was looking at this the other day and came up with a workaround.  Let me know if this works for you.  I suspect we will add better support for this as we continue work on the Operator App Framework.  We've used a method described in splunk-ansible for [generic Splunk config setting](https://github.com/splunk/splunk-ansible/blob/develop/docs/advanced/default.yml.spec.md#configuration-files) in the Docker container.  

Using the example @gburton1 mentioned above, bindDNpassword, I've created a file called value.txt.  Note there is a blank line on line 1 of value.txt.
````

splunk:
   conf:
   - key: authentication
     value: 
        directory: /opt/splunk/etc/system/local
        content:
          AD:
            bindDN : ldapuser.example.net
            bindDNpassword : foobar
````

Then create a secret out of this file
`kubectl create secret generic splunk-ldap-secret --from-file=default.yml=value.txt`

Then I started up a single instance Splunk pointing to this secret
````
cat <<EOF | kubectl apply -f -
apiVersion: enterprise.splunk.com/v2
kind: Standalone
metadata:
  name: s1
  finalizers:
  - enterprise.splunk.com/delete-pvc
spec:
  volumes:
    - name: default
      secret:
        secretName: splunk-ldap-secret
  defaultsUrl: /mnt/default/default.yml
EOF
````
When the container comes up in my k8s cluster
````
bash-3.2$ kubectl get pods
NAME                                  READY   STATUS    RESTARTS   AGE
splunk-default-monitoring-console-0   1/1     Running   0          16m
splunk-operator-6fc84675d7-29qn2      1/1     Running   0          7d1h
splunk-s1-standalone-0                1/1     Running   0          18m
````
I can go verify the settings have been acquired from the secret
````
bash-3.2$ kubectl exec splunk-s1-standalone-0 --stdin --tty $1 -- /bin/bash
[splunk@splunk-s1-standalone-0 splunk]$ cd etc/system/local
[splunk@splunk-s1-standalone-0 local]$ cat authentication.conf 

[AD]
bindDN = ldapuser.example.net
bindDNpassword = foobar
[splunk@splunk-s1-standalone-0 local]$ 
````

**Comment by akondur** [comment]
Hi @gburton1 and @locmai , please let us know if the workaround mentioned [above](https://github.com/splunk/splunk-operator/issues/657#issuecomment-1080967050) helps inject kubernetes secrets into the Splunk configuration files. As @pogdin mentioned we are looking into a solution for the same via AppFramework in the future.

**Comment by locmai** [comment]
Hi @akondur , we are about to test the approach from @pogdin in the (maybe) next sprint and will let you know the result asap. 

To me, it looks like what we need right now and it makes sense also.

**Comment by akondur** [comment]
Hi @locmai , I wanted to check in to see if you were able to test the approach?

**Comment by akondur** [comment]
Hi @locmai , we are closing the issue per the solution provided. Please re-open the issue if needed.

**Comment by gburton1** [comment]
Sorry we weren't able to get to this. We appreciate the solution provided, and we'll come back to this work in the future.

**Comment by marcusschiesser** [comment]
I wrote a [small shell script](https://marcusschiesser.de/2022/06/24/injecting-passwords-from-kubernetes-secrets-into-splunk-config-files/) to generate the secrets from config files like this:

```yaml
splunk:
   conf:
   - key: authentication
     value: 
        directory: /opt/splunk/etc/system/local
        content:
          AD:
            bindDNpassword : ${AD_PASSWORD}
```

That's helpful as these configuration can be quite large

**Comment by marcusschiesser** [comment]
@pogdin There seems to be a problem with the workaround: If I am mounting the `default.yml`, seems like the [Global K8S Secret object](https://splunk.github.io/splunk-operator/Examples.html#managing-global-kubernetes-secret-object) from the operator is not used.

**Comment by marcusschiesser** [comment]
@pogdin Great that you can verify the issue. But this means that #832 is a general bug, right?

**Comment by akondur** [comment]
Hi @marcusschiesser , closing this issue as #832 is closed now.

**Comment by marcusschiesser** [comment]
@akondur I would open this issue again as closing #832 still leaves this issue open. AFAIK it's only possible with the workaround above to configure passwords in config files

**Comment by akondur** [comment]
Re-opening the issue as we can provide an enhancement in the future for injecting passwords.

**Comment by akondur** [comment]
Per [comment](https://github.com/splunk/splunk-operator/issues/863#issuecomment-1444222029), closing as this is a linked issue. Please re-open if the parent issue still persists.

**Comment by marcusschiesser** [comment]
@akondur please reopen this issue is not fixed by #863

**Comment by akondur** [comment]
Hi @marcusschiesser , please open a new Github issue and mark it as an enhancement to allow for injecting passwords.
